### DISCUSSION TITLE
Version 0.16.1 no longer supports naming the binary as 'server'

### DISCUSSION THREAD

**Discussion by usrme** [pocketbase/pocketbase#2693] [Q&A] [answered]
I've been successfully running PocketBase on Ubuntu 18.04 (I know, end-of-life, etc.), but after upgrading to 0.16.1 the executable no longer starts as expected due to me using the name "server" for the executable and having the service's `WorkingDirectory` as the same directory where the executable is.

```console
/home/user/data/server serve --http 0.0.0.0:8090
-bash: /home/user/data/server: Is a directory
```

This is because upon starting the executable a `server` directory is created in the working directory. If I start the executable under `root` without a service, then it creates a `server` directory in the `root` user's home directory, which includes the `migrations` directory with a migration `.go` file.

I can easily change the executable's name on my end, but perhaps this could be addressed nonetheless.

**Comment by ganigeorgiev** [comment]
I'm not sure I follow.

We don't create a `server` directory. In your case I think the error implies that the path `/home/user/data/server` is not an executable but a directory (or maybe you don't have `+x` permission?).

**Reply by usrme** [reply]
This is indeed peculiar because I literally changed nothing in the deployment process besides upgrading PocketBase and then this happened. I would have no reason to add a `server` directory myself, let alone one containing the migrations, nor remove the execution bit.

**Reply by ganigeorgiev** [reply]
You can name the executable however you want as long as it doesn't collide with an existing entry in its parent directory.

There are not enough details about your setup to understand or reproduce the issue and I suspect that  it is caused by some misconfiguration in the deployment process.

**Reply by usrme** [reply]
I'll try to reproduce this in a container and show the steps, but I literally just deployed a brand new Ubuntu 18.04 VM, extracted a `.tar.gz` archive containing just the executable named `server` and a `README.md` file, and tried to start the following service:

```console
$ cat /etc/systemd/system/server.service
[Unit]
Description = Server service

[Service]
Type             = simple
User             = user
Group            = user
LimitNOFILE      = 4096
Restart          = always
RestartSec       = 5s
WorkingDirectory = /home/user/data
ExecStart        = /home/user/data/server serve --http 0.0.0.0:8090

[Install]
WantedBy = multi-user.target
```

And was greeted with the following _from_ the executable:

```console
$ sudo journalctl -u server.service -n 500
-- Logs begin at Fri 2022-03-04 07:39:26 UTC, end at Mon 2023-06-12 11:28:55 UTC. --
Jun 12 11:28:05 router systemd[1]: Started Server service.
Jun 12 11:28:09 router server[5385]: 2023/06/12 11:28:09 mkdir server: not a directory
...
```

And here are the contents of the `data` directory:

```console
$ ls -la data/
total 54476
drwxrwxr-x 3 user user     4096 Jun 12 11:27 .
drwxr-xr-x 7 user user     4096 Jun 12 11:27 ..
-rw-rw-r-- 1 user user     8068 Jun 12 11:21 README.md
-rw-rw-r-- 1 user user 13682695 Jun 12 11:27 server_0.0.27_linux_amd64.tar.gz
drwxrwxr-x 2 user user     4096 Jun 12 11:28 pb_data
-rwxr-xr-x 1 user user 42078208 Jun 12 11:24 server
```

If I switch back to 0.16.0 and try the exact same process on a brand new VM again, it will work.

**Reply by ganigeorgiev** [reply]
Based on the archive name you have a custom build, aka. using PocketBase as framework, right?

If that's the case, do you have `Automigrate` enabled by default? If yes, then it is possible that it will create a snapshot go migration on first `serve` but there are no differences in the behavior between v0.16.0 and v0.16.1.

To prevent the useless Go migration snapshot creation, I'll consider updating the migration examples to show using a conditional bool check for Automigrate (aka. to enable it only during development).

**Reply by usrme** [reply]
Yes, this is something I should have mentioned from the start: I'm using PocketBase as a framework with `Automigrate` set to `true`. 

It's really odd though that I've literally never ran into this on versions prior to 0.16.1, but any version after that and it's an automatic failure on the first try. The example above was me trying 0.16.5 and getting hit right away.

I would really appreciate an example on how to avoid this as I'm very much interested in keeping my version of PocketBase up-to-date :pray:

**Reply by ganigeorgiev** [reply, marked as answer]
You can disable the `Automigrate` by using a build flag, env variable or loosely check whether `go run` or `go build` was used. For example:

```go
// loosely check if it was run using "go run"
isGoRun := strings.HasPrefix(os.Args[0], os.TempDir())

// note that in the upcoming v0.17.0 release the declaration will slightly change
migratecmd.MustRegister(app, app.RootCmd, &migratecmd.Options{
	Automigrate: isGoRun,
})
```

I'll consider documenting it better or maybe adding a special case when `Automigrate` with Go migration templates are used.

**Reply by usrme** [reply]
I usually build releases through `goreleaser`, is there something that would work for that out of the box? If not, I'll just opt for the environment variable and we can consider this closed.

**Reply by ganigeorgiev** [reply]
Note that there isn't an easy and 100% reliable way to detect automatically whether you've started the executable with `go run` vs `go build` because every time when you execute `go run` under the hood Go compiles it in some temp directory and starts it (_that's why in my example I check `os.TempDir()`_).

I'm not sure if there is anything builtin in goreleaser, but usually your options for manual detections are:

- use an env variable and check it inside your Go code eg. with `os.GetEnv("IS_PROD")`

- use a global variable and override it with ldflags (similar to how we set the global package version when creating the prebuilt executable - https://github.com/pocketbase/pocketbase/blob/master/.goreleaser.yaml#L14)

- use a go build tag, eg. `go build -tags=prod` (you can check [Build constraints](https://pkg.go.dev/cmd/go#hdr-Build_constraints) and `go help build`).

**Reply by usrme** [reply]
Thank you so much!
